### PR TITLE
feat: Customizable email subject name

### PR DESCRIPTION
**Pull request by puridach-w**
### SUMMARY
This feature allows users to customize the email subject when creating a report or an alert within Superset. This is driven by a need to include specific parameters within the email subject line, enhancing the specificity and resonance of alert or report notifications. 

Changes:
- Implementation of a new text input field within the 'create alert/report' interface, through which users can specify their desired email subject line.
- If the text field is left blank, a default email subject line will be generated by the system.
- The user interface will show the new text field solely when the email sending method is selected by a user.

This feature has been tested with Superset master branch.
This proposed feature has already been discussed in the Superset Improvement Proposal (see https://github.com/apache/superset/issues/25945)

### BEFORE SCREENSHOTS
image

image

### AFTER SCREENSHOTS
image

image

### TESTING INSTRUCTIONS
To verify the changes made, please follow these steps: (Assume that there TAG for alert and report and open)
- Go to the Alert and report creation page
- Create one of them
- Fill all the information that required
- Select Notification method as "Email"
- You will see input box for "email subject name"
- Fill the desired subject name and save it
- Wait for email send and check the email subject name

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by puridach-w** [comment]
Run: superset db downgrade

Screenshot 2566-12-21 at 12 53 24

Run: superset db upgrade

Screenshot 2566-12-21 at 12 53 35

File: superset/reports/notifications/email.py
Comment (giftig) [inline superset/reports/notifications/email.py:181]: Feels like this could still be useful even with per-schedule customisation, as you might want to prefix all your emails with a standard prefix and then choose individual subject lines after the prefix. I'd continue supporting it, though we could probably improve the behaviour where prefix is `None` or empty by removing any leading space etc.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (giftig) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:423]: "name" doesn't seem right here. Subject line, or subject text, or just subject, makes more sense I think.

Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx]: I already resolve this one it is unused code ad I already remove it thanks

File: superset/reports/notifications/email.py
Comment (puridach-w) [reply superset/reports/notifications/email.py:181]: I think if the user wants to have prefix then they can always add it themselves and we should allow user 100% control over their email subject.

For the default subject line it still have prefix I just modify code to this

```
if self._report_schedule.chart:
                name = (
                    f"{prefix} "
                    f"{self._report_schedule.name}: "
                    f"{self._report_schedule.chart.slice_name}"
                )
            else:
                name = (
                    f"{prefix} "
                    f"{self._report_schedule.name}: "
                    f"{self._report_schedule.dashboard.dashboard_title}"
                )
```

**Comment by puridach-w** [comment]
> ## [Codecov](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache) Report
> Attention: `6 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`8c32c6d`)](https://app.codecov.io/gh/apache/superset/commit/8c32c6da169afec312923e516850d90a69e78f46?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache) 69.18% compared to head [(`f7c7a9a`)](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache) 59.04%.
> > Report is 5 commits behind head on master.
> 
> [Files](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache)	Patch %	Lines
> [superset/commands/report/execute.py](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache#diff-c3VwZXJzZXQvY29tbWFuZHMvcmVwb3J0L2V4ZWN1dGUucHk=)	0.00%	[6 Missing ⚠️ ](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/apache/superset/pull/26327?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=apache).

What should I do with this? Can you suggest please.

**Comment by john-bodley** [comment]
ping @eschutho and @lilykuang who are working on [SIP-110](https://github.com/apache/superset/issues/26183).

**Comment by puridach-w** [comment]
@giftig @john-bodley Could you please help me review this PR again?

**Comment by giftig** [comment]
@puridach-w the code all looks ok to me but I'm not very familiar with the subject of reports. I think it'd be a good idea to get a review from @eschutho or @lilykuang since they're working on the related SIP John mentioned above; would be good to confirm what's done here aligns ok with that work.

File: superset/migrations/versions/2024-01-03_16-57_61304d31c995_add_subject_column_to_report_schedule.py
Comment (eschutho) [inline superset/migrations/versions/2024-01-03_16-57_61304d31c995_add_subject_column_to_report_schedule.py:35]: I think a `String(length=255)` type should be all you need here, rather than a text string?

File: superset/reports/models.py
Comment (eschutho) [inline superset/reports/models.py:173]: Same here `String(255)`

File: superset/reports/notifications/email.py
Comment (eschutho) [reply superset/reports/notifications/email.py:181]: To keep backward compatibility, can we continue to include the prefix if `app.config["EMAIL_REPORTS_SUBJECT_PREFIX"]` exists or is not blank/empty/nullish? This way each application can decide for themselves if they would like to include a prefix or not.

File: superset/reports/schemas.py
Comment (eschutho) [inline superset/reports/schemas.py:153]: I think you want to move `example` and description into a metadata dictionary. You can also define a new description variable for the email_subject.

File: superset/reports/schemas.py
Comment (eschutho) [inline superset/reports/schemas.py:285]: same as above

File: superset/commands/report/execute.py
Comment (eschutho) [inline superset/commands/report/execute.py:395]: This name variable is really just for internal logs. I'm not sure if the subject line would help here?

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (eschutho) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: instead of using `onInputChange` here, you can just create a new event handler like this for example:
```
  const onOwnersChange = (value: Array<SelectValue>) => {
    updateAlertState('owners', value || []);
  };
  ```

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (eschutho) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:931]: Instead of updating the subject line here, you can also add the formatting when saving.

File: superset-frontend/src/features/alerts/components/NotificationMethod.tsx
Comment (eschutho) [inline superset-frontend/src/features/alerts/components/NotificationMethod.tsx:131]: since you're passing the email subject down as a prop, you can use the prop value from the parent instead of storing it here.

**Review by eschutho — DISMISSED** [review]
I left a few comments. Feel free to ping me if you'd like any clarification on anything or have questions.

File: superset-frontend/src/features/alerts/components/NotificationMethod.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/components/NotificationMethod.tsx:131]: Modified

File: superset/migrations/versions/2024-01-03_16-57_61304d31c995_add_subject_column_to_report_schedule.py
Comment (puridach-w) [reply superset/migrations/versions/2024-01-03_16-57_61304d31c995_add_subject_column_to_report_schedule.py:35]: You're right I already modified it

File: superset/reports/models.py
Comment (puridach-w) [reply superset/reports/models.py:173]: Modified

File: superset/reports/schemas.py
Comment (puridach-w) [reply superset/reports/schemas.py:153]: Modified

File: superset/reports/schemas.py
Comment (puridach-w) [reply superset/reports/schemas.py:285]: Modified

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: But it also needs to update the 'report name' value in the state before updating the email subject. That is the reason why I added this code. It is used to show the default subject line immediately when changing the report name. If a separate function is used, code must also be added to set the new value into the state, which will cause redundant code.
```
if (name === 'name') {
      updateEmailSubject();
    }
```

File: superset/commands/report/execute.py
Comment (puridach-w) [reply superset/commands/report/execute.py:395]: If i didn't misunderstand the code that use for send alert is smtp
`send_email_smtp(
                to,
                subject,
                content.body,
                app.config,
                files=[],
                data=content.data,
                images=content.images,
                bcc="",
                mime_subtype="related",
                dryrun=False,
                header_data=content.header_data,
            )`
   
 and subject it come from this 
 `def _get_subject(self) -> str:
        return __(
            "%(title)s",
            title=self._content.name,
        )`

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx]: Modified

**Comment by puridach-w** [comment]
Ping @eschutho I have already modified the code based on your comment.

**Comment by puridach-w** [comment]
Ping @eschutho Could you please help me review this PR again?

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (eschutho) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:1179]: since the default is handled in execute.py (server side) I don't think you need to also define it here in the client. It can just be left blank.

File: superset/reports/notifications/email.py
Comment (eschutho) [reply superset/reports/notifications/email.py:181]: I agree with @giftig on the premise that the app.config is set by the application admin, whereas the title will be set by individual users. If an application admin set an app.config prefix, it's likely something important that they don't want removed from the emails, such as "Company Confidential", etc.

File: superset/commands/report/execute.py
Comment (eschutho) [reply superset/commands/report/execute.py:397]: Thanks for pointing that out. You're correct. Maybe just a small nit for readability, I would switch the logic to 
````
if self._report_schedule.email_subject:
    name = self._report_schedule.email_subject
else:
...
```
I also left a separate comment about persisting the app.config prefix.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1179]: As you can see in the image the default value will be shown if user doesn't input any text and this code is a part of that function.

image

File: superset/commands/report/execute.py
Comment (puridach-w) [reply superset/commands/report/execute.py]: Modified

File: superset/reports/notifications/email.py
Comment (puridach-w) [reply superset/reports/notifications/email.py]: Ok, I got it. I already modify to be as previous.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: First video: https://github.com/apache/superset/assets/76474644/47df84ce-e7c1-4db9-b008-3b43b6c39592

From the first video, it can be observed that the Report or Dashboard name changes whenever the _default value that's shown as a placeholder changes. The need for two variables arises from the requirement to display the default Subject name, and another one to store the customized name.

Second Video: https://github.com/apache/superset/assets/76474644/bb024b5e-96e6-4323-ada7-cef981752341

As shown in the second video, initially when I customize the "Email subject name", it changes again when I edit the report name. This is caused by using only the currentAlert.subject_name as the placeholder and value for the "Email subject name" field. Therefore, when the report name is changed, the "Email subject name" field gets updated.

Do you think it's important to display the default email subject line?

**Comment by puridach-w** [comment]
Ping @eschutho Could you please help me review this PR again?

**Comment by rusackas** [comment]
A rebase of this branch should fix the CI issue you're seeing. 🤞

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: @eschutho would you kindly provide your opinion on this matter?

Comment (eschutho) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: Thanks for the videos. I'll defer to a product or design person on the functionality. @yousoph do you think the default subject line should appear in the email subject field?

Comment (eschutho) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:896]: Sorry for the delayed response here, but the rest is looking good to me. Thanks for all the contributions here @puridach-w!

**Comment by yousoph** [comment]
Hi @puridach-w ! Thanks for this new feature, it's looking good 

About the subject line UI, I think we can make it a required field in the form, but filled out with the default report name when the form is opened and it can update as the user changes the inputs. If the user clears it they will see that the field is required. There should be patterns on the modal that you can look to for behavior of required fields (red asterisk for example), but let me know if there's anything you'd like help defining! 

Thanks!

**Comment by geido** [comment]
Hello @puridach-w, feel free to ping me when you are ready for another round of review. Thanks!

**Comment by puridach-w** [comment]
> http://18.246.236.152:8080



> /testenv up FEATURE_ALERT_REPORTS=true

FYI, I still haven't edited the code from yousoph comment. But I appreciate your care and attention. Thank you very much.

**Comment by geido** [comment]
Np @puridach-w. There are a few conflicts as I just merged a PR that changes the styles of the A&R modal. I hope that does not become too much of a hassle. Let me know!

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (fisjac) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1250]: Perhaps the placeholder could show the default if empty? This would also allow for users to revert back to default if they've saved a report with a name, but want to return to the Superset default subject line.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (rtexelm) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:985]: I think this is causing [the valid report unit test](https://github.com/apache/superset/blob/d6dbb95cc45360ce24778811ab5a6d5f9a4198e6/superset-frontend/src/features/alerts/AlertReportModal.test.tsx#L255) to fail. Might be causing your CI error. Do you need to push `email_subject` to `settings` in `onNotificationAdd`?

File: superset-frontend/src/features/alerts/AlertReportModal.tsx
Comment (puridach-w) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1250]: Ping @eschutho @yousoph, What are your thoughts on reverting to displaying the default value in the placeholder? This way, users can easily revert to the default if they accidentally delete the email_subject. For example, if we make the "email_subject" field a required field and the user deletes it, it could be harder for them to restore the default value.

Comment (geido) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1250]: I think it makes sense

Comment (yousoph) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1250]: Sounds good to me

**Comment by puridach-w** [comment]
Ping @eschutho @geido I request another review of my PR? Thank you!

**Comment by geido** [comment]
Hey @puridach-w doing some manual testing. It seems that the spacing is off here

Screenshot 2024-03-11 at 16 41 35

**Comment by puridach-w** [comment]
> Hey @puridach-w if you look at the testenv above, you will see that there is a` :` in the placeholder. I think something might be broken in the placeholder logics.
> 
> Screenshot 2024-03-19 at 19 44 18

Hi @geido , I've updated the placeholder logic, but now there are some failing checks in the [test-postgres-hive]. I haven't made changes to that area could you help me with the problem?

**Comment by puridach-w** [comment]
Ping @geido Could you please help me review this PR again?

**Comment by geido** [comment]
@puridach-w not sure if this is related to your PR or not but I cannot add a report. It thinks recipients are not entered even though I did. Can you verify that, please?

**Comment by puridach-w** [comment]
> @puridach-w not sure if this is related to your PR or not but I cannot add a report. It thinks recipients are not entered even though I did. Can you verify that, please? 

Sure let me check

**Comment by puridach-w** [comment]
Hi @geido , as I checked, I found that the issue come from the updateNotificationSetting function in the AlertReportModal file. The problem arises because the function does not update the state with the new setting value, resulting in the recipient variable doesn't have a value. I've fixed this by adding an else statement to update the setting value.

**Comment by geido** [comment]
Hey @puridach-w this is looking much better now. However, I noticed that you can add an empty space as a valid value for the subject. Should we probably strip it or allow for a minimum amount of chars? Wdyt?

**Comment by puridach-w** [comment]
> Hey @puridach-w this is looking much better now. However, I noticed that you can add an empty space as a valid value for the subject. Should we probably strip it or allow for a minimum amount of chars? Wdyt?

Hi @geido. I agree and have added the condition to trap that case. Could you please review it again? There's a failed job under 'Build & Publish Docker Images / docker-build', but I noticed there is an open PR aimed at resolving the issue.

**Review by geido — APPROVED** [review]
Thanks for all the hard work on this!

**Pull request merged**
